mrjob/spark/mrjob_spark_harness.py

```diff
@@ -279,20 +292,23 @@ def combiner_helper(pairs1, pairs2):
     return rdd
 
 
-def _shuffle_and_sort(rdd, sort_values=False):
+def _shuffle_and_sort(rdd, sort_values=False, num_reducers=None):
     """Simulate Hadoop's shuffle-and-sort step, so that data will be in the
     format the reducer expects.
 
     :param rdd: an RDD containing lines representing encoded key-value pairs,
                 where the encoded key comes first and is followed by a TAB
                 character (the encoded key may not contain TAB).
-    :sort_values: if true, ensure all lines corresponding to a given key
-                  are sorted (by their encoded value)
+    :param sort_values: if true, ensure all lines corresponding to a given key
+                        are sorted (by their encoded value)
+    :param num_reducers: limit the number of paratitions of output rdd, which
+                         is similar to mrjob's limit on number of reducers.
     :return: an RDD containing "reducer ready" lines representing encoded
              key-value pairs, that is, where all lines with the same key are
              adjacent and in the same partition
     """
-    rdd = rdd.groupBy(lambda line: line.split(b'\t')[0])
+    rdd = rdd.groupBy(
+        lambda line: line.split(b'\t')[0], numPartitions=num_reducers)
```

Comment:
For this, and `combineByKey()`, once we set a specific number of partitions, we're pretty much committed to no re-partitions until the end of the step. I'd expect to see `preservesPartitioning=bool(num_reducers)` on every operation that follows this one.

Once you do this, you'd probably want to add a test case or two to `tests.spark.test_mrjob_spark_harness.PreservesPartitioningTestCase`.